### PR TITLE
PIM-6031: Deleted families are still visible on the product grid

### PR DESCRIPTION
**Pull request by anaelChardan**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md) 

**Description (for Contributor and Core Developer)**

As the family is stored directly on the product in mongoDb, we have to delete it when the family itself is deleted :)

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | :white_check_mark:
| Added Behats                      | :white_check_mark:
| Changelog updated                 | :white_check_mark:
| Review and 2 GTM                  | :clock1:
| Micro Demo to the PO (Story only) |  :negative_squared_cross_mark:
| Migration script                  |  :negative_squared_cross_mark: 
| Tech Doc                          |  :negative_squared_cross_mark:

:white_check_mark: Done and pass

:red_circle: Done but fail

:clock1: Done but pending

:negative_squared_cross_mark: Not needed

### CODE REVIEW

**Comment by a2xchip** [comment]
PIM-6034 in commits, why PR is 6031?

**Comment by anaelChardan** [comment]
@a2xchip yes, sorry it is a mistake, corrected :-)

**Comment by a2xchip** [comment]
@anaelChardan are commits wrong? cause PR is still 31

**Comment by anaelChardan** [comment]
@a2xchip nop, it was just the commit name that was wrong, I mixed two tickets in my head ^^, the ticket is PIM-6031 :-)

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FamilyDeletedQueryGenerator extends AbstractQueryGenerator
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function generateQuery($entity, $field, $oldValue, $newValue)
+    {
```
Comment (a2xchip) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php]: What is in field?

**Comment by a2xchip** [comment]
@anaelChardan That happens sometimes :-)

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php (hunk 1 above)
Comment (anaelChardan) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php]: It could be code, I don't know, let me check :)

Comment (anaelChardan) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php]: So, In my opinion "field" should be empty, do not hesitate to correct me if I'm wrong as I'm a new on the product :-), thanks for you help

Comment (anaelChardan) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php]: Nice, thanks for the review :-)

File: features/family/delete_family.feature [hunk 2]
```diff
@@ -10,24 +10,32 @@ Feature: Delete a family
 
   Scenario: Successfully delete a family from the grid
     Given I am on the families page
-    Then I should see family boots
+    And I should see family boots
```
Comment (ClementGautier) [inline features/family/delete_family.feature]: `I should see` is an assertion, so, as it follow a "given" state, it should use a "Then" statement.
In the other hand, if this step does no assertion, it should not use the auxiliary "should".
Pick one but don't mix them ;)

File: features/family/delete_family.feature [hunk 3]
```diff
@@ -10,24 +10,32 @@ Feature: Delete a family
...
+    Then I should be on the families page
+    And the grid should contain 4 elements
+    But I should not see family sneakers
 
   Scenario: Successfully delete a family used by a product
     Given the following product:
       | sku | family   |
       | foo | sneakers |
+    And I am on the products page
+    And I display the columns SKU, Family
```
Comment (ClementGautier) [inline features/family/delete_family.feature]: You do an action here: use `When`.

File: features/family/delete_family.feature [hunk 4]
```diff
@@ -10,24 +10,32 @@ Feature: Delete a family
...
+    Then I should see the text "sneakers"
     And I edit the "sneakers" family
```
Comment (ClementGautier) [inline features/family/delete_family.feature]: change this `And` to a `When` as you switch from an assertion to an action again.

File: features/family/delete_family.feature [hunk 5]
```diff
@@ -10,24 +10,32 @@ Feature: Delete a family
...
     When I press the "Delete" button
```
Comment (ClementGautier) [inline features/family/delete_family.feature]: this `When` will become a `And`

File: features/family/delete_family.feature [hunk 6]
```diff
@@ -10,24 +10,32 @@ Feature: Delete a family
...
 
   Scenario: Successfully delete a family from the edit page
     Given I edit the "sneakers" family
     When I press the "Delete" button
     And I confirm the deletion
-    Then the grid should contain 4 elements
-    And I should not see family sneakers
+    Then I should be on the families page
+    And the grid should contain 4 elements
+    But I should not see family sneakers
```
Comment (damien-carcel) [inline features/family/delete_family.feature]: Add a "@jira"

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        return [
+            [
+                ['normalizedData.family.code' => $entity->getCode()],
+                ['$unset'   => ['normalizedData.family' => '']],
```
Comment (ClementGautier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGenerator.php]: don't align here (or align everywhere, not both)

File: src/Pim/Bundle/CatalogBundle/spec/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGeneratorSpec.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        $this->beConstructedWith($namingUtility, Family::class);
+    }
+
+    public function it_generates_a_query_to_delete_product_family(FamilyInterface $family)
+    {
+        $family->getCode()->willReturn(42);
+        $this->generateQuery($family, '', '', '')->shouldReturn([
+            [
+                ['normalizedData.family.code' => 42],
+                ['$unset'   => ['normalizedData.family' => '']],
```
Comment (ClementGautier) [inline src/Pim/Bundle/CatalogBundle/spec/Doctrine/MongoDBODM/QueryGenerator/FamilyDeletedQueryGeneratorSpec.php]: same, align or not, don't mix ;)

File: features/family/delete_family.feature (hunk 4 above)
Comment (anaelChardan) [inline features/family/delete_family.feature]: Yes

File: features/family/delete_family.feature (hunk 3 above)
Comment (anaelChardan) [inline features/family/delete_family.feature]: Yes :-)

File: features/family/delete_family.feature (hunk 2 above)
Comment (anaelChardan) [inline features/family/delete_family.feature]: @damien-carcel asks me to do this, but I agree :D

**Pull request merged**